runtime/artifact.go

```diff
@@ -0,0 +1,123 @@
+package runtime
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	got "github.com/taskcluster/go-got"
+	"github.com/taskcluster/taskcluster-client-go/queue"
+	"github.com/taskcluster/taskcluster-client-go/tcclient"
+	"github.com/taskcluster/taskcluster-worker/runtime/ioext"
+)
+
+// TODO: Consider making Name, TaskID, RunID be in a BaseArtifact?
+type S3Artifact struct {
+	Name     string
+	Mimetype string
+	Expires  tcclient.Time
+	Stream   ioext.ReadSeekCloser
+}
+
+type ErrorArtifact struct {
+	Name    string
+	Message string
+	Reason  string
+	Expires tcclient.Time
+}
+
+type RedirectArtifact struct {
+	Name     string
+	MimeType string
+	URL      string
+	Expires  tcclient.Time
+}
+
+// UploadS3Artifact is responsible for creating new artifacts
+// in the queue and then performing the upload to s3.
+// TODO: More docs here.
+func UploadS3Artifact(artifact S3Artifact, context *TaskContext) error {
+	req, err := json.Marshal(queue.S3ArtifactRequest{
+		ContentType: artifact.Mimetype,
+		Expires:     artifact.Expires,
+		StorageType: "s3",
+	})
+	par := queue.PostArtifactRequest(req)
+	parsp, _, err := context.Queue().CreateArtifact(
+		context.TaskID,
+		strconv.Itoa(context.RunID),
+		artifact.Name,
+		&par,
+	)
+	if err != nil {
+		// TODO: Do something with all of these errors
+	}
+	var resp queue.S3ArtifactResponse
+	err = json.Unmarshal(json.RawMessage(*parsp), &resp)
+	if err != nil {
+		// TODO: Do something with all of these errors
+	}
+
+	putArtifact(resp.PutURL, artifact.Mimetype, artifact.Stream)
+
+	return err
+}
+
+func putArtifact(urlStr, mime string, stream ioext.ReadSeekCloser) error {
+	// TODO: Make this do retries
+	// TODO: Use https://golang.org/pkg/bufio/
+	u, err := url.Parse(urlStr)
+	contentLength, err := stream.Seek(0, 2)
+
+	header := make(http.Header)
+	header.Set("content-type", mime)
+
+	// TODO: Mock out http client
+	backoff := got.DefaultBackOff
+	attempts := 0
+	client := &http.Client{}
+	for {
+		attempts++
+		stream.Seek(0, 0)
+		req := &http.Request{
+			Method:        "PUT",
+			URL:           u,
+			Proto:         "HTTP/1.1",
+			ProtoMajor:    1,
+			ProtoMinor:    1,
+			Header:        header,
+			Body:          stream,
+			ContentLength: contentLength,
+		}
+		resp, err := client.Do(req)
```

Comment:
@imbstack, this is fine short term...

@gregarndt, @imbstack, long term I think we need `TaskContext` to implement the `Context` interface from:
https://godoc.org/golang.org/x/net/context

Then we can use: https://godoc.org/golang.org/x/net/context/ctxhttp
Here and ensure that the http request stops uploading if the task is called, or some other plugin returns an error while we are uploading artifacts... then the plugin manager can cancel...(okay for plugin manager to cancel it would have to wrap it somehow)...

So for plugin manager to cancel we would have to pass a Context object with every request. And this would have to be a separate Context object, not just TaskContext... 

So maybe:
1) TaskContext implmenets the Context interface
2) Every hooks on plugin takes a Context interface as first parameter

Then pluginManager will call `plugin.Prepare(Context, TaskContext)`, and `plugin.Stopped(Context, ResultSet)`, where Context is TaskContext.WithCancel()